stl/inc/atomic

```diff
@@ -354,8 +369,18 @@ struct _Atomic_storage {
         _Check_memory_order(_Order);
         const auto _Storage_ptr  = _STD addressof(_Storage);
         const auto _Expected_ptr = _STD addressof(_Expected);
+#if _CMPXCHG_MASK_OUT_PADDING_BITS
+        if constexpr (_Cmpxchg_has_padding_bits_v<_Ty>) {
+            __builtin_zero_non_value_bits(_Expected_ptr);
+        }
+#endif // _CMPXCHG_MASK_OUT_PADDING_BITS
         bool _Result;
         _Lock();
+#if _CMPXCHG_MASK_OUT_PADDING_BITS
+        if constexpr (_Cmpxchg_has_padding_bits_v<_Ty>) {
+            __builtin_zero_non_value_bits(_Storage_ptr);
+        }
+#endif // _CMPXCHG_MASK_OUT_PADDING_BITS

```

Comment:
Hmmmm do you think it would be worth it to avoid writing to _Storage when the CAS fails to avoid making that cache line 'dirty' like this?

```
auto _Local = _Storage;
__builtin_zero_non_value_bits(_STD addressof(_Local));
if (_CSTD memcmp(_STD addressof(_Local), _Expected_ptr, sizeof(_Ty)) == 0) {
// as before
}
?
```

Reply:
Does CAS write to storage when it fails?
I thought about loading the current value before first CAS to copy padding bits from it, and did not see it as pure win:
- most o the times the expectation for CAS is either from previous CAS or from preceding load
- extra load delays attempt
- 64-bit load on x86 is not necessarily very efficient
- 128-bit load on x64 is definitely inefficient (it is another CAS)

Reply:
Ah  this is non-lock-free case. Then yes

Reply:
Fixed